### PR TITLE
Add ListBuilder::with_field to support non nullable list fields (#5330)

### LINKED ISSUE #5330: Support Creating Non-Nullable Lists with ListBuilder
**Describe the bug**
When I try to write a `non-null List<non-null i32>` to a parquet file, I get an error. Might be related to #385.

**To Reproduce**
```rust
use std::{error::Error, fs::File, sync::Arc};

use ::arrow::record_batch::RecordBatch;
use arrow::{
    array::{Int32Builder, ListBuilder},
    datatypes::{DataType, Field, Schema},
};

use parquet::arrow::arrow_writer::ArrowWriter;

pub fn main() -> Result<(), Box<dyn Error>> {
    let mut builder = ListBuilder::new(Int32Builder::new());

    builder.append_value([Some(1), Some(2), Some(3)]);

    let array = builder.finish();

    let inner_nullable = false;

    let schema = Arc::new(Schema::new(vec![Field::new_list(
        "col_a",
        Field::new_list_field(DataType::Int32, inner_nullable),
        false,
    )]));

    let batch = RecordBatch::try_new(schema, vec![Arc::new(array)])?;

    let buf = File::create("test.parquet")?;
    let mut writer = ArrowWriter::try_new(buf, batch.schema(), None)?;
    writer.write(&batch)?;
    writer.close()?;

    Ok(())
}
```

**Expected behavior**
I expect no errors, but instead I get:

```
Error: InvalidArgumentError("column types must match schema types, 

expected
List(Field { name: \"item\", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} })

but found 
List(Field { name: \"item\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })

at column index 0")

```

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #5330

# Rationale for this change
 

This lets users override the `Field` used by `ListBuilder`, allowing overriding the nullability, name, and other metadata. It also switches over to using the safer `ListArray` constructors.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/builder/generic_list_builder.rs [hunk 1]
```diff
@@ -116,6 +116,19 @@ impl<OffsetSize: OffsetSizeTrait, T: ArrayBuilder> GenericListBuilder<OffsetSize
...
             values_builder,
+            field: None,
+        }
+    }
+
+    /// Override the field passed to [`GenericListArray::new`]
+    ///
+    /// By default a nullable field is created with the name `item`
+    ///
+    /// Note: [`Self::finish`] and [`Self::finish_cloned`] will panic if the provided data type does not match `T`
```
Comment (tustvold) [inline arrow-array/src/builder/generic_list_builder.rs:127]: The ergonomics of this aren't ideal, but aside from adding a `data_type` method to `ArrayBuilder` I'm not sure of a way around this.

File: arrow-array/src/builder/generic_list_builder.rs [hunk 2]
```diff
@@ -116,6 +116,20 @@ impl<OffsetSize: OffsetSizeTrait, T: ArrayBuilder> GenericListBuilder<OffsetSize
             offsets_builder,
             null_buffer_builder: NullBufferBuilder::new(capacity),
             values_builder,
+            field: None,
+        }
+    }
+
+    /// Override the field passed to [`GenericListArray::new`]
```
Comment (tustvold) [inline arrow-array/src/builder/generic_list_builder.rs:123]: This avoids duplicating the docs about what the implications of different nullability options are

File: arrow-array/src/builder/generic_list_builder.rs [hunk 3]
```diff
@@ -275,53 +289,37 @@ where
...
-        let array_data_builder = ArrayData::builder(data_type)
-            .len(len)
-            .add_buffer(offset_buffer)
-            .add_child_data(values_data)
-            .nulls(null_bit_buffer);
 
-        let array_data = unsafe { array_data_builder.build_unchecked() };
+        let field = match &self.field {
+            Some(f) => f.clone(),
+            None => Arc::new(Field::new("item", values.data_type().clone(), true)),
```
Comment (dvic) [inline arrow-array/src/builder/generic_list_builder.rs:302]:
```suggestion
            None => Arc::new(Field::new_list_field(values.data_type().clone(), true)),
```

Comment (tustvold) [reply arrow-array/src/builder/generic_list_builder.rs:302]: This would create a field of `DataType::List` which isn't what we want here

Comment (dvic) [reply arrow-array/src/builder/generic_list_builder.rs:302]: Ahh, could it be that you've mistaken `new_list_field` for `new_list`? https://docs.rs/arrow-schema/latest/arrow_schema/struct.Field.html#method.new_list

Comment (tustvold) [reply arrow-array/src/builder/generic_list_builder.rs:302]: Oh I didn't realise the field version got added in the end 😅. I'm old fashioned and prefer the explicit form, but good suggestion

Comment (alamb) [reply arrow-array/src/builder/generic_list_builder.rs:302]: Yeah, I think there are two camps in terms of "do I know / want to know what the name of the anonymous field are in the list"

Specifically, 
1. if you are familiar with the convention,  then it is better to have the code explcit. 
2. If you don't know the `"item"` convention having a function hide it is easier to understand

File: arrow-array/src/builder/generic_list_builder.rs [hunk 4]
```diff
@@ -765,4 +763,39 @@ mod tests {
         assert_eq!(0, i1.null_count());
         assert_eq!(i1.values(), &[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
     }
+
+    #[test]
+    fn test_non_nullable_list() {
+        let field = Arc::new(Field::new("item", DataType::Int32, false));
```
Comment (alamb) [inline arrow-array/src/builder/generic_list_builder.rs:769]: Would it make sense to also add a test for a field that is not named "item"?

**Review by alamb — APPROVED** [review]
LGTM - thank you @tustvold  and @dvic

File: arrow-array/src/builder/generic_list_builder.rs [hunk 5]
```diff
@@ -765,8 +765,8 @@ mod tests {
     }
 
     #[test]
-    fn test_non_nullable_list() {
-        let field = Arc::new(Field::new("item", DataType::Int32, false));
+    fn test_with_field() {
+        let field = Arc::new(Field::new("bar", DataType::Int32, false));
```
Comment (alamb) [inline arrow-array/src/builder/generic_list_builder.rs:769]: 👍

**Pull request merged**